File: src-tauri/src/discovery/subagent_locator.rs
/// Locate subagent JSONL files for a session.
///
/// Supports two directory structures:
/// NEW: {projectId}/{sessionId}/subagents/agent_*.jsonl
/// OLD: {projectId}/agent_*.jsonl

use std::path::Path;

pub fn has_subagents(projects_dir: &Path, project_id: &str, session_id: &str) -> bool {
    let base_dir = super::path_decoder::extract_base_dir(project_id);

    // Check new structure first
    let new_path = projects_dir
        .join(base_dir)
        .join(session_id)
        .join("subagents");
    if new_path.is_dir() {
        if let Ok(entries) = std::fs::read_dir(&new_path) {
            for entry in entries.flatten() {
                let name = entry.file_name();
                let name = name.to_string_lossy();
                if name.ends_with(".jsonl") {
                    return true;
                }
            }
        }
    }

    // Check old structure: agent_*.jsonl at project root
    let project_dir = projects_dir.join(base_dir);
    if let Ok(entries) = std::fs::read_dir(&project_dir) {
        for entry in entries.flatten() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            if name.starts_with("agent_") && name.ends_with(".jsonl") {
                return true;
            }
        }
    }

    false
}

File: internal/discovery/subagent_locator.go
package discovery

// Subagent locator — find subagent JSONL files for a session.
//
// Supports two layouts:
//
//	NEW: {projectDir}/{sessionID}/subagents/*.jsonl
//	OLD: {projectDir}/agent_*.jsonl

import (
	"os"
	"path/filepath"
	"strings"
)

// SubagentPath builds the NEW-layout path for a single subagent JSONL file.
// Mirrors subagent_locator::subagent_path.
func SubagentPath(projectsDir, projectID, parentSessionID, subagentID string) string {
	baseDir := ExtractBaseDir(projectID)
	return filepath.Join(projectsDir, baseDir, parentSessionID, "subagents", subagentID+".jsonl")
}

// HasSubagents returns true if any subagent JSONL file exists for the session.
// Mirrors subagent_locator::has_subagents.
func HasSubagents(projectsDir, projectID, sessionID string) bool {
	baseDir := ExtractBaseDir(projectID)

	// Check new structure first: {baseDir}/{sessionID}/subagents/
	newPath := filepath.Join(projectsDir, baseDir, sessionID, "subagents")
	if info, err := os.Stat(newPath); err == nil && info.IsDir() {
		entries, err := os.ReadDir(newPath)
		if err == nil {
			for _, e := range entries {
				if strings.HasSuffix(e.Name(), ".jsonl") {
					return true
				}
			}
		}
	}

	// Check old structure: {baseDir}/agent_*.jsonl
	projectDir := filepath.Join(projectsDir, baseDir)
	entries, err := os.ReadDir(projectDir)
	if err != nil {
		return false
	}
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, "agent_") && strings.HasSuffix(name, ".jsonl") {
			return true
		}
	}
	return false
}
